server/igc/IgcHelper.ts

```diff
@@ -0,0 +1,182 @@
+  for (let i = 0; i < fixes.length; i++) {
+    if (i > 0) {
+      const deltaTimestamp = fixes[i].timestamp - fixes[i - 1].timestamp;
+      if (deltaTimestamp > 0) {
+        fixes[i].hspeed = calculateSpeed(
+          fixes[i],
+          fixes[i - 1],
+          deltaTimestamp / 1000
+        );
+
+        const alt1 = fixes[i].gpsAltitude;
+        const alt2 = fixes[i - 1].gpsAltitude;
+
+        if (alt1 && alt2) {
+          const deltaAltitude = alt1 - alt2;
+
+          fixes[i].vspeed = (deltaAltitude / deltaTimestamp) * 1000;
+        } else {
+          fixes[i].vspeed = 0;
+        }
+      } else {
+        fixes[i].hspeed = fixes[i - 1].hspeed;
+        fixes[i].vspeed = fixes[i - 1].vspeed;
+      }
+    } else {
+      fixes[i].hspeed = 0;
+      fixes[i].vspeed = 0;
+    }
+  }
+
+  for (let i = 0; i < fixes.length; i++) {
+    const now = fixes[i].timestamp;
+    let start, end;
+    for (
+      start = i;
+      start > 0 &&
+      fixes[start].timestamp > now - Math.round((maPeriod * 1000) / 2);
+      start--
+    );
+    for (
+      end = i;
+      end < fixes.length - 1 &&
+      fixes[end].timestamp < now + Math.round((maPeriod * 1000) / 2);
+      end++
+    );
+    const maSegment = fixes.slice(start, end + 1);
+
+    fixes[i].hma =
+      maSegment.reduce(
+        (sum, x) => (fixes[i].hma = x.hspeed ? sum + x.hspeed : 0),
+        0
+      ) / maSegment.length;
+    fixes[i].vma =
+      maSegment.reduce((sum, x) => sum + Math.abs(x.vspeed ?? 0), 0) /
+      maSegment.length;
+  }
+}
+
+function detectFlight(fixes: ExtendedBRecord[]) {
+  let start: number | undefined;
+  for (let i = 0; i < fixes.length - 1; i++) {
+    if (
+      start === undefined &&
+      fixes[i].hma! > definitionFlight.xt &&
+      fixes[i].vma! > definitionFlight.zt
+    )
+      start = i;
+    if (start !== undefined)
+      if (
+        fixes[i].hma! > definitionFlight.x0 &&
+        fixes[i].vma! > definitionFlight.z0
+      ) {
+        if (
+          fixes[i].timestamp >
+          fixes[start].timestamp + definitionFlight.t * 1000
+        )
+          for (let j = start; j <= i; j++) fixes[i].stateFlight = true;
+      } else {
+        start = undefined;
+      }
+  }
+}
+
+function detectGround(fixes: ExtendedBRecord[]) {
+  let start;
+  for (let i = 0; i < fixes.length - 1; i++) {
+    if (
+      start === undefined &&
+      fixes[i].hma! < definitionGround.xmax &&
+      fixes[i].vma! < definitionGround.zmax
+    )
+      start = i;
+    if (start !== undefined)
+      if (
+        fixes[i].hma! < definitionGround.xmax &&
+        fixes[i].vma! < definitionGround.zmax
+      ) {
+        if (
+          fixes[i].timestamp >
+          fixes[start].timestamp + definitionGround.t * 1000
+        )
+          for (let j = start; j <= i; j++) fixes[i].stateGround = true;
+      } else {
+        start = undefined;
+      }
+  }
+}
+
+export function detectLaunchLanding(fixes: ExtendedBRecord[]) {
+  let ll = [];
```

Comment:
ll is the short for???

Reply:
launch & landing (I didn't name it like that. It comes from the original implementation.)